GCC build seems to be previously broken:
```
[ 87%] Building CXX object Interop/PInvoke/SetLastError/CMakeFiles/SetLastErrorNative.dir/SetLastErrorNative.cpp.o
/__w/1/s/src/tests/profiler/native/eltprofiler/slowpatheltprofiler.cpp:30:82: error: ‘visibility’ attribute ignored [-Werror=attributes]
   30 | PROFILER_STUB EnterStub(FunctionIDOrClientID functionId, COR_PRF_ELT_INFO eltInfo)
```

See https://github.com/dotnet/runtime/pull/60225#issuecomment-948874503

--- a/src/coreclr/jit/redundantbranchopts.cpp
+++ b/src/coreclr/jit/redundantbranchopts.cpp
@@ -149,8 +149,10 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
                 const bool domIsSameRelop = matchCmp || matchSwp;
                 const bool domIsRevRelop  = matchRev || matchSwpRev;
 
-                // Note we could also infer the tree relop's value from similar relops higher in the dom tree.
-                // For example, (x >= 0) dominating (x > 0), or (x < 0) dominating (x > 0).
+                // Note we could also infer the tree relop's value from relops higher in the dom tree
+                // that involve the same operands but are not swaps or reverses.
+                //
+                // For example, (x >= 0) dominating (x > 0).
                 //
                 // That is left as a future enhancement.
                 //

--- a/src/coreclr/jit/redundantbranchopts.cpp
+++ b/src/coreclr/jit/redundantbranchopts.cpp
@@ -128,26 +128,40 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
 
             if (domCmpTree->OperKind() & GTK_RELOP)
             {
-                // We can use liberal VNs as bounds checks are not yet
+                // We can use liberal VNs here, as bounds checks are not yet
                 // manifest explicitly as relops.
                 //
-                ValueNum treeVN      = tree->GetVN(VNK_Liberal);
-                ValueNum domCmpVN    = domCmpTree->GetVN(VNK_Liberal);
-                ValueNum domCmpRevVN = vnStore->GetReversedRelopVN(domCmpVN);
+                // Look for an exact match and also try the various swapped/reversed forms.
+                //
+                const ValueNum treeVN   = tree->GetVN(VNK_Liberal);
+                const ValueNum domCmpVN = domCmpTree->GetVN(VNK_Liberal);
+                const ValueNum domCmpSwpVN =
+                    vnStore->GetRelatedRelop(domCmpVN, ValueNumStore::VN_RELATION_KIND::VRK_Swap);
+                const ValueNum domCmpRevVN =
+                    vnStore->GetRelatedRelop(domCmpVN, ValueNumStore::VN_RELATION_KIND::VRK_Reverse);
+                const ValueNum domCmpSwpRevVN =
+                    vnStore->GetRelatedRelop(domCmpVN, ValueNumStore::VN_RELATION_KIND::VRK_SwapReverse);
+
+                const bool matchCmp       = ((domCmpVN != ValueNumStore::NoVN) && (domCmpVN == treeVN));
+                const bool matchSwp       = ((domCmpSwpVN != ValueNumStore::NoVN) && (domCmpSwpVN == treeVN));
+                const bool matchRev       = ((domCmpRevVN != ValueNumStore::NoVN) && (domCmpRevVN == treeVN));
+                const bool matchSwpRev    = ((domCmpSwpRevVN != ValueNumStore::NoVN) && (domCmpSwpRevVN == treeVN));
+                const bool domIsSameRelop = matchCmp || matchSwp;
+                const bool domIsRevRelop  = matchRev || matchSwpRev;
 
                 // Note we could also infer the tree relop's value from similar relops higher in the dom tree.
                 // For example, (x >= 0) dominating (x > 0), or (x < 0) dominating (x > 0).
                 //
                 // That is left as a future enhancement.
                 //
-                if ((domCmpVN == treeVN) || ((domCmpRevVN != ValueNumStore::NoVN) && (domCmpRevVN == treeVN)))
+                if (domIsSameRelop || domIsRevRelop)
                 {
                     // The compare in "tree" is redundant.
                     // Is there a unique path from the dominating compare?
                     //
-                    const bool domIsSameRelop = (domCmpVN == treeVN);
-                    JITDUMP("\nDominator " FMT_BB " of " FMT_BB " has relop with %s liberal VN\n", domBlock->bbNum,
-                            block->bbNum, domIsSameRelop ? "the same" : "a reverse");
+                    JITDUMP("\nDominator " FMT_BB " of " FMT_BB " has %srelop with %s liberal VN\n", domBlock->bbNum,
+                            block->bbNum, matchSwp || matchSwpRev ? "swapped " : "",
+                            domIsSameRelop ? "the same" : "a reverse");
                     DISPTREE(domCmpTree);
                     JITDUMP(" Redundant compare; current relop:\n");
                     DISPTREE(tree);

--- a/src/coreclr/jit/redundantbranchopts.cpp
+++ b/src/coreclr/jit/redundantbranchopts.cpp
@@ -131,6 +131,7 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
                 // We can use liberal VNs as bounds checks are not yet
                 // manifest explicitly as relops.
                 //
+                ValueNum treeVN      = tree->GetVN(VNK_Liberal);
                 ValueNum domCmpVN    = domCmpTree->GetVN(VNK_Liberal);
                 ValueNum domCmpRevVN = vnStore->GetReversedRelopVN(domCmpVN);
 
@@ -139,13 +140,12 @@ bool Compiler::optRedundantBranch(BasicBlock* const block)
                 //
                 // That is left as a future enhancement.
                 //
-                if ((domCmpVN == tree->GetVN(VNK_Liberal)) ||
-                    ((domCmpRevVN != ValueNumStore::NoVN) && (domCmpRevVN == tree->GetVN(VNK_Liberal))))
+                if ((domCmpVN == treeVN) || ((domCmpRevVN != ValueNumStore::NoVN) && (domCmpRevVN == treeVN)))
                 {
                     // The compare in "tree" is redundant.
                     // Is there a unique path from the dominating compare?
                     //
-                    const bool domIsSameRelop = domCmpVN == tree->GetVN(VNK_Liberal);
+                    const bool domIsSameRelop = (domCmpVN == treeVN);
                     JITDUMP("\nDominator " FMT_BB " of " FMT_BB " has relop with %s liberal VN\n", domBlock->bbNum,
                             block->bbNum, domIsSameRelop ? "the same" : "a reverse");
                     DISPTREE(domCmpTree);

--- a/src/coreclr/jit/valuenum.cpp
+++ b/src/coreclr/jit/valuenum.cpp
@@ -4451,21 +4451,24 @@ bool ValueNumStore::IsVNHandle(ValueNum vn)
 }
 
 //------------------------------------------------------------------------
-// GetReversedRelopVN: return value number for reversed comparison
+// GetRelatedRelop: return value number for reversed/swapped comparison
 //
 // Arguments:
-//    vn - vn to reverse
+//    vn - vn to base things on
+//    vnk - whether the new vn should swap, reverse, or both
 //
 // Returns:
-//    vn for reversed comparsion, or NoVN.
+//    vn for reversed/swapped comparsion, or NoVN.
 //
 // Note:
-//    If "vn" corresponds to (x > y)
-//    then reverse("vn") corresponds to (x <= y)
+//    If "vn" corresponds to (x > y), the resulting VN correponds to
+//    VRK_Swap               (y < x)
+//    VRK_Reverse            (x <= y)
+//    VRK_SwapReverse        (y >= x)
 //
 //    Will return NoVN for all float comparisons.
 //
-ValueNum ValueNumStore::GetReversedRelopVN(ValueNum vn)
+ValueNum ValueNumStore::GetRelatedRelop(ValueNum vn, VN_RELATION_KIND vrk)
 {
     if (vn == NoVN)
     {
@@ -4498,46 +4501,108 @@ ValueNum ValueNumStore::GetReversedRelopVN(ValueNum vn)
         return NoVN;
     }
 
-    // Map to the reverse VNFunc
-    //
-    VNFunc reverseOp;
+    const bool reverse = (vrk == VN_RELATION_KIND::VRK_Reverse) || (vrk == VN_RELATION_KIND::VRK_SwapReverse);
+    const bool swap    = (vrk == VN_RELATION_KIND::VRK_Swap) || (vrk == VN_RELATION_KIND::VRK_SwapReverse);
 
-    if (funcAttr.m_func >= VNF_Boundary)
+    // Set up the new function
+    //
+    VNFunc newFunc = funcAttr.m_func;
+
+    // Swap the predicate, if so asked.
+    //
+    if (swap)
     {
-        switch (funcAttr.m_func)
+        if (newFunc >= VNF_Boundary)
         {
-            case VNF_LT_UN:
-                reverseOp = VNF_GE_UN;
-                break;
-            case VNF_LE_UN:
-                reverseOp = VNF_GT_UN;
-                break;
-            case VNF_GE_UN:
-                reverseOp = VNF_LT_UN;
-                break;
-            case VNF_GT_UN:
-                reverseOp = VNF_LE_UN;
-                break;
-            default:
+            switch (newFunc)
+            {
+                case VNF_LT_UN:
+                    newFunc = VNF_GT_UN;
+                    break;
+                case VNF_LE_UN:
+                    newFunc = VNF_GE_UN;
+                    break;
+                case VNF_GE_UN:
+                    newFunc = VNF_LE_UN;
+                    break;
+                case VNF_GT_UN:
+                    newFunc = VNF_LT_UN;
+                    break;
+                default:
+                    return NoVN;
+            }
+        }
+        else
+        {
+            const genTreeOps op = (genTreeOps)newFunc;
+
+            if (!GenTree::OperIsRelop(op))
+            {
                 return NoVN;
+            }
+
+            newFunc = (VNFunc)GenTree::SwapRelop(op);
         }
     }
-    else
-    {
-        const genTreeOps op = (genTreeOps)funcAttr.m_func;
 
-        if (!GenTree::OperIsRelop(op))
-        {
-            return NoVN;
-        }
-
-        reverseOp = (VNFunc)GenTree::ReverseRelop(op);
-    }
-
-    // Create the resulting VN
+    // Reverse the predicate, if so asked.
     //
-    ValueNum reverseValueVN = VNForFunc(TYP_INT, reverseOp, funcAttr.m_args[0], funcAttr.m_args[1]);
-    return VNWithExc(reverseValueVN, excepVN);
+    if (reverse)
+    {
+        if (newFunc >= VNF_Boundary)
+        {
+            switch (newFunc)
+            {
+                case VNF_LT_UN:
+                    newFunc = VNF_GE_UN;
+                    break;
+                case VNF_LE_UN:
+                    newFunc = VNF_GT_UN;
+                    break;
+                case VNF_GE_UN:
+                    newFunc = VNF_LT_UN;
+                    break;
+                case VNF_GT_UN:
+                    newFunc = VNF_LE_UN;
+                    break;
+                default:
+                    return NoVN;
+            }
+        }
+        else
+        {
+            const genTreeOps op = (genTreeOps)newFunc;
+
+            if (!GenTree::OperIsRelop(op))
+            {
+                return NoVN;
+            }
+
+            newFunc = (VNFunc)GenTree::ReverseRelop(op);
+        }
+    }
+
+    // Create the resulting VN, swapping arguments if needed.
+    //
+    ValueNum newVN  = VNForFunc(TYP_INT, newFunc, funcAttr.m_args[swap ? 1 : 0], funcAttr.m_args[swap ? 0 : 1]);
+    ValueNum result = VNWithExc(newVN, excepVN);
+
+#ifdef DEBUG
+    if (m_pComp->verbose)
+    {
+        printf("%s of ", swap ? (reverse ? "swap-reverse" : "swap") : "reverse");
+        m_pComp->vnPrint(vn, 1);
+        printf(" => ");
+        m_pComp->vnPrint(newVN, 1);
+        if (result != newVN)
+        {
+            m_pComp->vnPrint(result, 1);
+        }
+        printf("\n");
+    }
+#endif
+
+    return result;
 }
 
 bool ValueNumStore::IsVNConstantBound(ValueNum vn)

--- a/src/coreclr/jit/valuenum.cpp
+++ b/src/coreclr/jit/valuenum.cpp
@@ -4451,7 +4451,7 @@ bool ValueNumStore::IsVNHandle(ValueNum vn)
 }
 
 //------------------------------------------------------------------------
-// GetReversedRelopVN: return value number for reversed comparsion
+// GetReversedRelopVN: return value number for reversed comparison
 //
 // Arguments:
 //    vn - vn to reverse
@@ -4460,8 +4460,10 @@ bool ValueNumStore::IsVNHandle(ValueNum vn)
 //    vn for reversed comparsion, or NoVN.
 //
 // Note:
-//    If "vn" corresponds to (x > y) 
+//    If "vn" corresponds to (x > y)
 //    then reverse("vn") corresponds to (x <= y)
+//
+//    Will return NoVN for all float comparisons.
 //
 ValueNum ValueNumStore::GetReversedRelopVN(ValueNum vn)
 {
@@ -4485,6 +4487,13 @@ ValueNum ValueNumStore::GetReversedRelopVN(ValueNum vn)
     }
 
     if (funcAttr.m_arity != 2)
+    {
+        return NoVN;
+    }
+
+    // Don't try and model float compares.
+    //
+    if (varTypeIsFloating(TypeOfVN(funcAttr.m_args[0])))
     {
         return NoVN;
     }

--- a/src/coreclr/jit/valuenum.cpp
+++ b/src/coreclr/jit/valuenum.cpp
@@ -4455,13 +4455,13 @@ bool ValueNumStore::IsVNHandle(ValueNum vn)
 //
 // Arguments:
 //    vn - vn to base things on
-//    vnk - whether the new vn should swap, reverse, or both
+//    vrk - whether the new vn should swap, reverse, or both
 //
 // Returns:
 //    vn for reversed/swapped comparsion, or NoVN.
 //
 // Note:
-//    If "vn" corresponds to (x > y), the resulting VN correponds to
+//    If "vn" corresponds to (x > y), the resulting VN corresponds to
 //    VRK_Swap               (y < x)
 //    VRK_Reverse            (x <= y)
 //    VRK_SwapReverse        (y >= x)

--- a/src/coreclr/jit/valuenum.h
+++ b/src/coreclr/jit/valuenum.h
@@ -842,11 +842,18 @@ public:
     // Returns true iff the VN represents a handle constant.
     bool IsVNHandle(ValueNum vn);
 
-    // Given a relop VN, get the VN for the "reversed" relop, which has the opposite value
-    // eg given VN(x < y), return  VN(y <=x)
+    // Given VN(x > y), return VN(y > x), VN(x <= y) or VN(y >= x)
     //
     // If vn is not a relop, return NoVN.
-    ValueNum GetReversedRelopVN(ValueNum vn);
+    //
+    enum class VN_RELATION_KIND
+    {
+        VRK_Swap,       // (y >  x)
+        VRK_Reverse,    // (x <= y)
+        VRK_SwapReverse // (y >= x)
+    };
+
+    ValueNum GetRelatedRelop(ValueNum vn, VN_RELATION_KIND vrk);
 
     // Convert a vartype_t to the value number's storage type for that vartype_t.
     // For example, ValueNum of type TYP_LONG are stored in a map of INT64 variables.

